### ISSUE TITLE
Build errors in Ubuntu 14.04 and HHVM 3.7.1

### ISSUE THREAD

**Issue by foolishneo** [PocketRent/hhvm-pgsql#92]
Hi, 

I got the following errors when I was compiling the package:

[ 25%] Building CXX object CMakeFiles/pgsql.dir/pgsql.cpp.o
/home/Downloads/hhvm-pgsql/pgsql.cpp: In static member function ‘static HPHP::{anonymous}::PGSQL* HPHP::{anonymous}::PGSQL::Get(const HPHP::Variant&)’:
/home/Downloads/hhvm-pgsql/pgsql.cpp:210:69: error: request for member ‘get’ in ‘HPHP::Variant::toResource() const().HPHP::Resource::getTyped<HPHP::{anonymous}::PGSQL>(1, 1)’, which is of pointer type ‘HPHP::{anonymous}::PGSQL*’ (maybe you meant to use ‘->’ ?)
     PGSQL *pgsql = conn_id.toResource().getTyped<PGSQL>(true, true).get();
                                                                     ^
/home/Downloads/hhvm-pgsql/pgsql.cpp: In static member function ‘static HPHP::{anonymous}::PGSQLResult* HPHP::{anonymous}::PGSQLResult::Get(const HPHP::Variant&)’:
/home/Downloads/hhvm-pgsql/pgsql.cpp:306:71: error: request for member ‘get’ in ‘HPHP::Variant::toResource() const().HPHP::Resource::getTyped<HPHP::{anonymous}::PGSQLResult>(1, 1)’, which is of pointer type ‘HPHP::{anonymous}::PGSQLResult*’ (maybe you meant to use ‘->’ ?)
     auto *res = result.toResource().getTyped<PGSQLResult>(true, true).get();
                                                                       ^
/home/Downloads/hhvm-pgsql/pgsql.cpp: In member function ‘virtual void HPHP::{anonymous}::pgsqlExtension::moduleLoad(const Map&, HPHP::Hdf)’:
/home/Downloads/hhvm-pgsql/pgsql.cpp:1659:90: error: no matching function for call to ‘HPHP::Config::GetBool(const Map&, HPHP::Hdf&, const char [16], bool)’
         PGSQL::AllowPersistent     = Config::GetBool(ini,  pgsql, "AllowPersistent", true);
                                                                                          ^
/home/Downloads/hhvm-pgsql/pgsql.cpp:1659:90: note: candidates are:
In file included from /usr/include/hphp/runtime/base/runtime-option.h:30:0,
                 from /usr/include/hphp/runtime/base/memory-manager.h:34,
                 from /usr/include/hphp/runtime/base/array-data.h:25,
                 from /usr/include/hphp/runtime/base/array-data-defs.h:20,
                 from /usr/include/hphp/runtime/base/array-iterator.h:23,
                 from /home/Downloads/hhvm-pgsql/pgsql.cpp:4:
/usr/include/hphp/runtime/base/config.h:183:15: note: static bool HPHP::Config::GetBool(const IniSettingMap&, const HPHP::Hdf&, bool)
   static bool GetBool(const IniSettingMap &ini, const Hdf& config,
               ^
/usr/include/hphp/runtime/base/config.h:183:15: note:   candidate expects 3 arguments, 4 provided
/usr/include/hphp/runtime/base/config.h:221:15: note: static bool HPHP::Config::GetBool(const IniSettingMap&, const string&, bool)
   static bool GetBool(const IniSettingMap &ini, const std::string &name,
               ^
/usr/include/hphp/runtime/base/config.h:221:15: note:   candidate expects 3 arguments, 4 provided
/home/Downloads/hhvm-pgsql/pgsql.cpp:1660:86: error: no matching function for call to ‘HPHP::Config::GetInt32(const Map&, HPHP::Hdf&, const char [14], int)’
         PGSQL::MaxPersistent       = Config::GetInt32(ini, pgsql, "MaxPersistent", -1);
                                                                                      ^
/home/Downloads/hhvm-pgsql/pgsql.cpp:1660:86: note: candidates are:
In file included from /usr/include/hphp/runtime/base/runtime-option.h:30:0,
                 from /usr/include/hphp/runtime/base/memory-manager.h:34,
                 from /usr/include/hphp/runtime/base/array-data.h:25,
                 from /usr/include/hphp/runtime/base/array-data-defs.h:20,
                 from /usr/include/hphp/runtime/base/array-iterator.h:23,
                 from /home/Downloads/hhvm-pgsql/pgsql.cpp:4:
/usr/include/hphp/runtime/base/config.h:197:18: note: static int32_t HPHP::Config::GetInt32(const IniSettingMap&, const HPHP::Hdf&, int32_t)
   static int32_t GetInt32(const IniSettingMap &ini, const Hdf& config,
                  ^
/usr/include/hphp/runtime/base/config.h:197:18: note:   candidate expects 3 arguments, 4 provided
/usr/include/hphp/runtime/base/config.h:237:18: note: static int32_t HPHP::Config::GetInt32(const IniSettingMap&, const string&, int32_t)
   static int32_t GetInt32(const IniSettingMap &ini, const std::string &name,
                  ^
/usr/include/hphp/runtime/base/config.h:237:18: note:   candidate expects 3 arguments, 4 provided
/home/Downloads/hhvm-pgsql/pgsql.cpp:1661:81: error: no matching function for call to ‘HPHP::Config::GetInt32(const Map&, HPHP::Hdf&, const char [9], int)’
         PGSQL::MaxLinks            = Config::GetInt32(ini, pgsql, "MaxLinks", -1);
                                                                                 ^
/home/Downloads/hhvm-pgsql/pgsql.cpp:1661:81: note: candidates are:
In file included from /usr/include/hphp/runtime/base/runtime-option.h:30:0,
                 from /usr/include/hphp/runtime/base/memory-manager.h:34,
                 from /usr/include/hphp/runtime/base/array-data.h:25,
                 from /usr/include/hphp/runtime/base/array-data-defs.h:20,
                 from /usr/include/hphp/runtime/base/array-iterator.h:23,
                 from /home/Downloads/hhvm-pgsql/pgsql.cpp:4:
/usr/include/hphp/runtime/base/config.h:197:18: note: static int32_t HPHP::Config::GetInt32(const IniSettingMap&, const HPHP::Hdf&, int32_t)
   static int32_t GetInt32(const IniSettingMap &ini, const Hdf& config,
                  ^
/usr/include/hphp/runtime/base/config.h:197:18: note:   candidate expects 3 arguments, 4 provided
/usr/include/hphp/runtime/base/config.h:237:18: note: static int32_t HPHP::Config::GetInt32(const IniSettingMap&, const string&, int32_t)
   static int32_t GetInt32(const IniSettingMap &ini, const std::string &name,
                  ^
/usr/include/hphp/runtime/base/config.h:237:18: note:   candidate expects 3 arguments, 4 provided
/home/Downloads/hhvm-pgsql/pgsql.cpp: In static member function ‘static HPHP::{anonymous}::PGSQLResult* HPHP::{anonymous}::PGSQLResult::Get(const HPHP::Variant&)’:
/home/Downloads/hhvm-pgsql/pgsql.cpp:308:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
make[2]: *** [CMakeFiles/pgsql.dir/pgsql.cpp.o] Error 1
make[1]: *** [CMakeFiles/pgsql.dir/all] Error 2
make: *** [all] Error 2

Could you look into this please?

**Comment by rny** [comment]
```
[ 25%] Building CXX object CMakeFiles/pgsql.dir/pgsql.cpp.o
/root/hhvm-pgsql/pgsql.cpp: In static member function ‘static HPHP::{anonymous}::PGSQL* HPHP::{anonymous}::PGSQL::Get(const HPHP::Variant&)’:
/root/hhvm-pgsql/pgsql.cpp:210:69: error: request for member ‘get’ in ‘HPHP::Variant::toResource() const().HPHP::Resource::getTyped<HPHP::{anonymous}::PGSQL>(1, 1)’, which is of pointer type ‘HPHP::{anonymous}::PGSQL*’ (maybe you meant to use ‘->’ ?)
     PGSQL *pgsql = conn_id.toResource().getTyped<PGSQL>(true, true).get();
                                                                     ^
/root/hhvm-pgsql/pgsql.cpp: In static member function ‘static HPHP::{anonymous}::PGSQLResult* HPHP::{anonymous}::PGSQLResult::Get(const HPHP::Variant&)’:
/root/hhvm-pgsql/pgsql.cpp:306:71: error: request for member ‘get’ in ‘HPHP::Variant::toResource() const().HPHP::Resource::getTyped<HPHP::{anonymous}::PGSQLResult>(1, 1)’, which is of pointer type ‘HPHP::{anonymous}::PGSQLResult*’ (maybe you meant to use ‘->’ ?)
     auto *res = result.toResource().getTyped<PGSQLResult>(true, true).get();
                                                                       ^
/root/hhvm-pgsql/pgsql.cpp: In member function ‘virtual void HPHP::{anonymous}::pgsqlExtension::moduleLoad(const Map&, HPHP::Hdf)’:
/root/hhvm-pgsql/pgsql.cpp:1659:90: error: no matching function for call to ‘HPHP::Config::GetBool(const Map&, HPHP::Hdf&, const char [16], bool)’
         PGSQL::AllowPersistent     = Config::GetBool(ini,  pgsql, "AllowPersistent", true);
                                                                                          ^
/root/hhvm-pgsql/pgsql.cpp:1659:90: note: candidates are:
In file included from /usr/include/hphp/runtime/base/runtime-option.h:30:0,
                 from /usr/include/hphp/runtime/base/memory-manager.h:34,
                 from /usr/include/hphp/runtime/base/array-data.h:25,
                 from /usr/include/hphp/runtime/base/array-data-defs.h:20,
                 from /usr/include/hphp/runtime/base/array-iterator.h:23,
                 from /root/hhvm-pgsql/pgsql.cpp:4:
/usr/include/hphp/runtime/base/config.h:183:15: note: static bool HPHP::Config::GetBool(const IniSettingMap&, const HPHP::Hdf&, bool)
   static bool GetBool(const IniSettingMap &ini, const Hdf& config,
               ^
/usr/include/hphp/runtime/base/config.h:183:15: note:   candidate expects 3 arguments, 4 provided
/usr/include/hphp/runtime/base/config.h:221:15: note: static bool HPHP::Config::GetBool(const IniSettingMap&, const string&, bool)
   static bool GetBool(const IniSettingMap &ini, const std::string &name,
               ^
/usr/include/hphp/runtime/base/config.h:221:15: note:   candidate expects 3 arguments, 4 provided
/root/hhvm-pgsql/pgsql.cpp:1660:86: error: no matching function for call to ‘HPHP::Config::GetInt32(const Map&, HPHP::Hdf&, const char [14], int)’
         PGSQL::MaxPersistent       = Config::GetInt32(ini, pgsql, "MaxPersistent", -1);
                                                                                      ^
/root/hhvm-pgsql/pgsql.cpp:1660:86: note: candidates are:
In file included from /usr/include/hphp/runtime/base/runtime-option.h:30:0,
                 from /usr/include/hphp/runtime/base/memory-manager.h:34,
                 from /usr/include/hphp/runtime/base/array-data.h:25,
                 from /usr/include/hphp/runtime/base/array-data-defs.h:20,
                 from /usr/include/hphp/runtime/base/array-iterator.h:23,
                 from /root/hhvm-pgsql/pgsql.cpp:4:
/usr/include/hphp/runtime/base/config.h:197:18: note: static int32_t HPHP::Config::GetInt32(const IniSettingMap&, const HPHP::Hdf&, int32_t)
   static int32_t GetInt32(const IniSettingMap &ini, const Hdf& config,
                  ^
/usr/include/hphp/runtime/base/config.h:197:18: note:   candidate expects 3 arguments, 4 provided
/usr/include/hphp/runtime/base/config.h:237:18: note: static int32_t HPHP::Config::GetInt32(const IniSettingMap&, const string&, int32_t)
   static int32_t GetInt32(const IniSettingMap &ini, const std::string &name,
                  ^
/usr/include/hphp/runtime/base/config.h:237:18: note:   candidate expects 3 arguments, 4 provided
/root/hhvm-pgsql/pgsql.cpp:1661:81: error: no matching function for call to ‘HPHP::Config::GetInt32(const Map&, HPHP::Hdf&, const char [9], int)’
         PGSQL::MaxLinks            = Config::GetInt32(ini, pgsql, "MaxLinks", -1);
                                                                                 ^
/root/hhvm-pgsql/pgsql.cpp:1661:81: note: candidates are:
In file included from /usr/include/hphp/runtime/base/runtime-option.h:30:0,
                 from /usr/include/hphp/runtime/base/memory-manager.h:34,
                 from /usr/include/hphp/runtime/base/array-data.h:25,
                 from /usr/include/hphp/runtime/base/array-data-defs.h:20,
                 from /usr/include/hphp/runtime/base/array-iterator.h:23,
                 from /root/hhvm-pgsql/pgsql.cpp:4:
/usr/include/hphp/runtime/base/config.h:197:18: note: static int32_t HPHP::Config::GetInt32(const IniSettingMap&, const HPHP::Hdf&, int32_t)
   static int32_t GetInt32(const IniSettingMap &ini, const Hdf& config,
                  ^
/usr/include/hphp/runtime/base/config.h:197:18: note:   candidate expects 3 arguments, 4 provided
/usr/include/hphp/runtime/base/config.h:237:18: note: static int32_t HPHP::Config::GetInt32(const IniSettingMap&, const string&, int32_t)
   static int32_t GetInt32(const IniSettingMap &ini, const std::string &name,
                  ^
/usr/include/hphp/runtime/base/config.h:237:18: note:   candidate expects 3 arguments, 4 provided
/root/hhvm-pgsql/pgsql.cpp: In static member function ‘static HPHP::{anonymous}::PGSQLResult* HPHP::{anonymous}::PGSQLResult::Get(const HPHP::Variant&)’:
/root/hhvm-pgsql/pgsql.cpp:308:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
make[2]: *** [CMakeFiles/pgsql.dir/pgsql.cpp.o] Error 1
make[1]: *** [CMakeFiles/pgsql.dir/all] Error 2
make: *** [all] Error 2

```

**Comment by rny** [comment]
git checkout tags/3.7.0

then 
```
-- The C compiler identification is GNU 4.8.2
-- The CXX compiler identification is GNU 4.8.2
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Configuring for HHVM API version 20150212
-- pgSQL Include dir: /usr/include/postgresql
-- libpq library: /usr/lib/libpq.so
-- Configuring done
-- Generating done
-- Build files have been written to: /root/hhvm-pgsql
Scanning dependencies of target pgsql
[ 25%] Building CXX object CMakeFiles/pgsql.dir/pgsql.cpp.o
[ 50%] Building CXX object CMakeFiles/pgsql.dir/pdo_pgsql_statement.cpp.o
In file included from /root/hhvm-pgsql/pdo_pgsql_resource.h:4:0,
                 from /root/hhvm-pgsql/pdo_pgsql_statement.cpp:1:
/usr/include/hphp/runtime/ext/pdo_driver.h:242:5: warning: ‘virtual HPHP::SmartPtr<HPHP::PDOResource> HPHP::PDODriver::createResourceImpl(const sp_PDOConnection&)’ was hidden [-Woverloaded-virtual]
     createResourceImpl(const sp_PDOConnection& conn) = 0;
     ^
In file included from /root/hhvm-pgsql/pdo_pgsql_statement.cpp:4:0:
/root/hhvm-pgsql/pdo_pgsql.h:10:35: warning:   by ‘virtual HPHP::SmartPtr<HPHP::PDOResource> HPHP::PDOPgSql::createResourceImpl()’ [-Woverloaded-virtual]
     virtual SmartPtr<PDOResource> createResourceImpl() override;
                                   ^
[ 75%] Building CXX object CMakeFiles/pgsql.dir/pdo_pgsql_connection.cpp.o
In file included from /root/hhvm-pgsql/pdo_pgsql_connection.h:4:0,
                 from /root/hhvm-pgsql/pdo_pgsql_connection.cpp:1:
/usr/include/hphp/runtime/ext/pdo_driver.h:242:5: warning: ‘virtual HPHP::SmartPtr<HPHP::PDOResource> HPHP::PDODriver::createResourceImpl(const sp_PDOConnection&)’ was hidden [-Woverloaded-virtual]
     createResourceImpl(const sp_PDOConnection& conn) = 0;
     ^
In file included from /root/hhvm-pgsql/pdo_pgsql_connection.cpp:4:0:
/root/hhvm-pgsql/pdo_pgsql.h:10:35: warning:   by ‘virtual HPHP::SmartPtr<HPHP::PDOResource> HPHP::PDOPgSql::createResourceImpl()’ [-Woverloaded-virtual]
     virtual SmartPtr<PDOResource> createResourceImpl() override;
                                   ^
[100%] Building CXX object CMakeFiles/pgsql.dir/pdo_pgsql.cpp.o
In file included from /root/hhvm-pgsql/pdo_pgsql.h:5:0,
                 from /root/hhvm-pgsql/pdo_pgsql.cpp:1:
/usr/include/hphp/runtime/ext/pdo_driver.h:242:5: warning: ‘virtual HPHP::SmartPtr<HPHP::PDOResource> HPHP::PDODriver::createResourceImpl(const sp_PDOConnection&)’ was hidden [-Woverloaded-virtual]
     createResourceImpl(const sp_PDOConnection& conn) = 0;
     ^
In file included from /root/hhvm-pgsql/pdo_pgsql.cpp:1:0:
/root/hhvm-pgsql/pdo_pgsql.h:10:35: warning:   by ‘virtual HPHP::SmartPtr<HPHP::PDOResource> HPHP::PDOPgSql::createResourceImpl()’ [-Woverloaded-virtual]
     virtual SmartPtr<PDOResource> createResourceImpl() override;
                                   ^
/root/hhvm-pgsql/pdo_pgsql.cpp:8:17: error: cannot declare variable ‘HPHP::s_pgsql_driver’ to be of abstract type ‘HPHP::PDOPgSql’
 static PDOPgSql s_pgsql_driver;
                 ^
In file included from /root/hhvm-pgsql/pdo_pgsql.cpp:1:0:
/root/hhvm-pgsql/pdo_pgsql.h:8:8: note:   because the following virtual functions are pure within ‘HPHP::PDOPgSql’:
 struct PDOPgSql : public PDODriver {
        ^
In file included from /root/hhvm-pgsql/pdo_pgsql.h:5:0,
                 from /root/hhvm-pgsql/pdo_pgsql.cpp:1:
/usr/include/hphp/runtime/ext/pdo_driver.h:242:5: note:         virtual HPHP::SmartPtr<HPHP::PDOResource> HPHP::PDODriver::createResourceImpl(const sp_PDOConnection&)
     createResourceImpl(const sp_PDOConnection& conn) = 0;
     ^
make[2]: *** [CMakeFiles/pgsql.dir/pdo_pgsql.cpp.o] Error 1
make[1]: *** [CMakeFiles/pgsql.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by rny** [comment]
build failed but use bin release 3.7.0 works.

**Comment by ChristopheLecroq** [comment]
Hi,

Same compilation problem here. I've try to use your release extension (3.7.0) but my HHVM (3.7.1) keeps crashing on Debian Jessie. Am I alone ?

Thanks

**Comment by rny** [comment]
build 3.6.0 vs hhvm lts 3.6.3 also failed. and that is also the reason of #91 

```
ubuntu@ip-172-30-3-181:~/hhvm-pgsql$ cmake . && make
-- Configuring for HHVM API version 20150212
-- pgSQL Include dir: /usr/include/postgresql
-- libpq library: /usr/lib/libpq.so
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ubuntu/hhvm-pgsql
[ 25%] Building CXX object CMakeFiles/pgsql.dir/pdo_pgsql_statement.cpp.o
In file included from /home/ubuntu/hhvm-pgsql/pdo_pgsql_statement.cpp:4:0:
/home/ubuntu/hhvm-pgsql/pdo_pgsql.h:10:18: error: conflicting return type specified for ‘virtual HPHP::PDOResource* HPHP::PDOPgSql::createResourceImpl()’
     PDOResource* createResourceImpl() override;
                  ^
In file included from /home/ubuntu/hhvm-pgsql/pdo_pgsql_resource.h:4:0,
                 from /home/ubuntu/hhvm-pgsql/pdo_pgsql_statement.cpp:1:
/usr/include/hphp/runtime/ext/pdo_driver.h:240:33: error:   overriding ‘virtual HPHP::SmartPtr<HPHP::PDOResource> HPHP::PDODriver::createResourceImpl()’
   virtual SmartPtr<PDOResource> createResourceImpl() = 0;
                                 ^
/usr/include/hphp/runtime/ext/pdo_driver.h:242:5: warning: ‘virtual HPHP::SmartPtr<HPHP::PDOResource> HPHP::PDODriver::createResourceImpl(const sp_PDOConnection&)’ was hidden [-Woverloaded-virtual]
     createResourceImpl(const sp_PDOConnection& conn) = 0;
     ^
In file included from /home/ubuntu/hhvm-pgsql/pdo_pgsql_statement.cpp:4:0:
/home/ubuntu/hhvm-pgsql/pdo_pgsql.h:10:18: warning:   by ‘virtual HPHP::PDOResource* HPHP::PDOPgSql::createResourceImpl()’ [-Woverloaded-virtual]
     PDOResource* createResourceImpl() override;
                  ^
/home/ubuntu/hhvm-pgsql/pdo_pgsql_statement.cpp: In member function ‘bool HPHP::PDOPgSqlStatement::create(const HPHP::String&, const HPHP::Array&)’:
/home/ubuntu/hhvm-pgsql/pdo_pgsql_statement.cpp:62:55: error: could not convert ‘this’ from ‘HPHP::PDOPgSqlStatement* const’ to ‘HPHP::sp_PDOStatement {aka HPHP::SmartPtr<HPHP::PDOStatement>}’
             int ret = pdo_parse_params(this, sql, nsql);
                                                       ^
make[2]: *** [CMakeFiles/pgsql.dir/pdo_pgsql_statement.cpp.o] Error 1
make[1]: *** [CMakeFiles/pgsql.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by shanewright** [comment]
@ChristopheLecroq you're not alone - I get the same on Ubuntu 14.04 (pre-built package for 3.7.0 crashes when used on HHVM 3.7.1...the only ubuntu package version available).

I also get similar/same compilation errors trying to build hhvm_pgsql myself - both from the 3.2.0 release, and from latest master.

Pretty much this is my showstopper for HHVM/Hack atm (aside from PHPStorm support ofc :/)

**Comment by rny** [comment]
i solved this issue by install the nightly version of hhvm and compile the master version of hhvm-pgsql.

**Comment by TheRedDevil** [comment]
I followed this procedure step-to-step, and last command (sudo make) fails with this error

```
 fatal error: hphp/runtime/ext/pdo_driver.h: No such file or directory
```

Does anyone have and idea what could be the problem? My ubuntu version is 14.04.

**Comment by rny** [comment]
fix again for the latest hhvm 3.9 

```
sudo apt-get install software-properties-common
sudo apt-key adv --recv-keys --keyserver hkp://keyserver.ubuntu.com:80 0x5a16e7281be7a449
sudo add-apt-repository 'deb http://dl.hhvm.com/ubuntu trusty main' -y
sudo apt-get update && sudo apt-get install -y hhvm-nightly

sudo apt-get update && sudo apt-get install -y hhvm-dev-nightly libpq-dev
# git clone https://github.com/PocketRent/hhvm-pgsql.git
git clone https://github.com/rny/hhvm-pgsql.git
cd hhvm-pgsql 
git checkout symbol_fix
hphpize && cmake . && make
```

**Comment by bensagardoy** [comment]
Compilation works for me with the latest hhvm 3.9.1 (not nightly) with the symbol_fix branch of your forked repo. Thanks a lot rny !

**Comment by Aatch** [comment]
Fixed by 160fa5c
